### PR TITLE
Final project work, v0.4.

### PR DESCRIPTION
**Pull request by pswish**
1. Why is this change being made?  Adding some code for development and markdown practice
2. What has changed? Added mvn project, added dataset csv file, added a pseudo map-hash-map function, added some pseudo test functions, added block diagram (img) for architecture overview.
3. How was this tested? Test with man test and man package, commented out the test files to pass the testing for now
4. Anything else to discus? Yes, more work and development is needed to make this into a java program. Need to add some more reviewers
5. Any other documentation? Not at this time

### CODE REVIEW

**Comment by pswish** [comment]
> I will touch base with Sean, to see if they indeed want to team up. I will send them a spreadsheet.

Thanks. As discussed, he might be only focused on the changemakers, as per Paul's direction, but we can always welcome another to the project. Let me know if he wants to set up a time to discus the plan going forward.

**Comment by learner-long-life** [comment]
This is a good start overall. Can you add more description to your diagram? I'm not sure what the Map means above, whether it's a mapping app like a GPS or like a HashMap.

I think your next step will be deciding what question to ask from the data. I also think there are some possiblities to frame this as a graph problem, with nodes being the stops on your trip and edges weighted by the cost of the trip (driving time, gas, mileage). It could also be an optimization problem if we can write some linear equations describing the constraints of each trip. Let's discuss more in class.

**Comment by pswish** [comment]
> This is a good start overall. Can you add more description to your diagram? I'm not sure what the Map means above, whether it's a mapping app like a GPS or like a HashMap.
> 
> I think your next step will be deciding what question to ask from the data. I also think there are some possiblities to frame this as a graph problem, with nodes being the stops on your trip and edges weighted by the cost of the trip (driving time, gas, mileage). It could also be an optimization problem if we can write some linear equations describing the constraints of each trip. Let's discuss more in class.

The map on the diagram was just meant to represent the top level map in the Map of Hash maps. I'll modify the diagram to be more clear. Thanks for the feedback on using a graph as i am not set on using the data structure i have selected so far. As for gas usage, that data needs to be collected from my app and weather too but both are easy to backdate.

**Comment by pswish** [comment]
Added a new concept drawing 
https://github.com/TheEvergreenStateCollege/upper-division-cs/blob/pswish-nathanmcl/image.png

**Review by nathanMcL — COMMENTED** [review]
I really appreciate these visual sudo representation.

File: dsa-23au/java-dsa/pswish-natmcl/pswish-app/src/main/java/com/pswishcorp/app/Driver.java [hunk 1]
```diff
@@ -1,7 +1,51 @@
...
+    public Object getDistanceFromHome() throws Exception {
+        throw new Exception("Not yet implemented, v0.4");
+    }
+
+    public Object getDriverName(String driver) {
+        Map<String, String> drivers = new HashMap<>();
+        return drivers;
+    }
+
+    public Object getCarMake() throws Exception {
```
Comment (EvergreenTor) [inline dsa-23au/java-dsa/pswish-natmcl/pswish-app/src/main/java/com/pswishcorp/app/Driver.java:38]: what does throws exception mean? I never covered "throws" keyword and exception handling (I think it's called that)

**Review by EvergreenTor — COMMENTED** [review]
I like you have started here! I just asked a couple questions to better my knowledge of the code.

File: dsa-23au/java-dsa/pswish-natmcl/pswish-app/src/test/java/com/pswishcorp/app/DriverTest.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
+// package com.pswishcorp.app;
+
+// import org.junit.Before;
+// import org.junit.Test;
+// import static org.junit.Assert.assertEquals;
```
[the remark this answers is not in the document]
Comment (pswish) [reply dsa-23au/java-dsa/pswish-natmcl/pswish-app/src/test/java/com/pswishcorp/app/DriverTest.java:5]: Good question. Nothing at the moment and none of the testing was working when i tried to compile (with mvn test)

File: dsa-23au/java-dsa/pswish-natmcl/pswish-app/src/main/java/com/pswishcorp/app/Driver.java (hunk 1 above)
Comment (pswish) [reply dsa-23au/java-dsa/pswish-natmcl/pswish-app/src/main/java/com/pswishcorp/app/Driver.java:38]: 'throws exception' causes the compile to spit out an error when compiling and is meant to be a placeholder and a reminder to update the code. Paul mentioned in class to change it to a runtime exception.

File: dsa-23au/java-dsa/pswish-natmcl/pswish-app/src/main/java/com/pswishcorp/app/Car.java [hunk 3]
```diff
@@ -4,16 +4,37 @@ public class Car {
...
 
     public static void main(String[] args) {
-        Car car1 = new Car("Paul","Q5", 27);
-        Car car2 = new Car("Nate", "Genesis", 28);
+        Car car1 = new Car("Paul","Suv", 27, .16);
+        Car car2 = new Car("Nate", "Sedan", 28,  .17);
         }
+
+    public Object getDriver() throws Exception {
+        throw new Exception("Not yet implemented, v0.4");
```
Comment (faulkdf) [inline dsa-23au/java-dsa/pswish-natmcl/pswish-app/src/main/java/com/pswishcorp/app/Car.java:25]: I am not super familiar with the 'throw' function. what are its uses/advantages?

File: image.png
Comment (faulkdf) [inline image.png:1]: What tool did you use to create this flowchart? this is a wonderful way to visualise data and I might do the same with my project.

Comment (pswish) [reply image.png:1]: Thanks! I used draw.io but I think that is internal to Amazon. There is another tool that very similar: https://app.diagrams.net/
Just click decide later to play around with it

File: dsa-23au/java-dsa/pswish-natmcl/pswish-app/src/main/java/com/pswishcorp/app/Driver.java (hunk 1 above)
Comment (pswish) [reply dsa-23au/java-dsa/pswish-natmcl/pswish-app/src/main/java/com/pswishcorp/app/Driver.java:38]: Here is some documentation: https://docs.oracle.com/javase/tutorial/essential/exceptions/throwing.html but think of it like you can use it as reminder to fix it later that should let you compile. My way isn’t the best way so I am looking to make changes soon.

File: image.png
Comment (faulkdf) [reply image.png:1]: Oh amazing! Thank you. It looks like the link you sent just redirects to draw.io so it must not just be internal to amazon.

**Pull request merged**